src/darwin/frida-helper-backend-glue.m

```diff
@@ -813,6 +814,292 @@ static gboolean frida_agent_context_init_functions (FridaAgentContext * self, Gu
+create_capstone (GumCpuType cpu_type, GumAddress start) {
+  csh capstone;
+  cs_err err;
+  switch (cpu_type)
+  {
+    case GUM_CPU_ARM64:
+      err = cs_open (CS_ARCH_ARM64, CS_MODE_LITTLE_ENDIAN, &capstone);
+      break;
+
+    case GUM_CPU_ARM:
+      if (start & 1)
+        err = cs_open (CS_ARCH_ARM, CS_MODE_THUMB, &capstone);
+      else
+        err = cs_open (CS_ARCH_ARM, CS_MODE_ARM, &capstone);
+      break;
+
+    case GUM_CPU_IA32:
+      err = cs_open (CS_ARCH_X86, CS_MODE_32, &capstone);
+      break;
+
+    case GUM_CPU_AMD64:
+      err = cs_open (CS_ARCH_X86, CS_MODE_64, &capstone);
+      break;
+
+    default:
+      g_assert_not_reached ();
+  }
+  g_assert_cmpint (err, ==, CS_ERR_OK);
+  return capstone;
+}
+
+static GumAddress
+find_libSystem_initializer_end (guint task, GumCpuType cpu_type,
+        GumAddress start, gsize max_size)
+{
+  uint64_t address = start & (~1);
+  csh capstone;
+  const uint8_t *code;
+  cs_err err;
+  GumAddress found = 0;
+  gpointer image;
+  cs_insn * insn = NULL;
+  size_t size;
+
+  capstone = create_capstone (cpu_type, start);
+
+  err = cs_option (capstone, CS_OPT_DETAIL, CS_OPT_ON);
+  g_assert_cmpint (err, ==, CS_ERR_OK);
+
+  image = gum_darwin_read (task, address, max_size, NULL);
+
+  insn = cs_malloc (capstone);
+  code = image;
+  size = max_size;
+
+  switch (cpu_type)
+  {
+    case GUM_CPU_ARM64:
+      while(cs_disasm_iter(capstone, &code, &size, &address, insn))
+      {
+        if (insn->id == ARM64_INS_RET)
+        {
+          found = insn->address;
+          break;
+        }
+      }
+      break;
+
+    case GUM_CPU_ARM:
+      {
+        int i, pop_lr = -1;
+        while(cs_disasm_iter(capstone, &code, &size, &address, insn))
+        {
+          if (insn->id == ARM_INS_PUSH &&
+              insn->address == (start & (~1)))
+          {
+            for (i=0; i != insn->detail->arm.op_count; i++)
+            {
+              if (insn->detail->arm.operands[i].reg == ARM_REG_LR)
+              {
+                pop_lr = i;
+                break;
+              }
+            }
+          }
+
+          if ((insn->id == ARM_INS_BX || insn->id == ARM_INS_BXJ) &&
+              insn->detail->arm.operands[0].type == ARM_OP_REG &&
+              insn->detail->arm.operands[0].reg == ARM_REG_LR)
+          {
+            found = insn->address;
+            break;
+          }
+
+          if (insn->id == ARM_INS_POP &&
+              pop_lr >= 0 &&
+              pop_lr < insn->detail->arm.op_count)
+          {
+            if (insn->detail->arm.operands[pop_lr].reg == ARM_REG_PC)
+            {
+              found = insn->address;
+              break;
+            }
+          }
+        }
+      }
+      break;
+
+    case GUM_CPU_IA32:
+    case GUM_CPU_AMD64:
+      while(cs_disasm_iter(capstone, &code, &size, &address, insn))
+      {
+        if (insn->id == X86_INS_RET  ||
+            insn->id == X86_INS_RETF ||
+            insn->id == X86_INS_RETFQ)
```

Comment:
uhm didn't do a `switch` here because it would have swallowed my `break`, requiring to make a longer condition in the while